### ISSUE TITLE
Ldap config question

### ISSUE THREAD

**Issue by reyman** [Galexrt/docker-zulip#62]
There is something i don't understand in option probably, i activate the `ZULIP_AUTH_BACKENDS: "ZulipLDAPAuthBackend"` and i use `SETTING_AUTH_LDAP_USER_SEARCH: "LDAPSearch(\"ou=Users,dc=comnmodel,dc=org\",ldap.SCOPE_SUBTREE,\"(uid=%(user)s)\")"` to find the correct user in LDAP. 

So, i don't use the email to login.

When i look to login page, there is only email / password login ... so how can i log correctly ? It seems the actual login page don't correspond to Ldap login page, isn't it ?

**Comment by Galexrt** [comment]
Can you start the Zulip container and `docker exec -it /bin/bash` and post the contents of the `/etc/zulip/settings.py` and the `/home/zulip/deployments/current/zproject/settings.py` (Remove your sensitive data).
When I have the configs I can look into your issue. :)

**Comment by reyman** [comment]
You can find the two gist here : 

https://gist.github.com/06ca03d5762f1ba67174

https://gist.github.com/e0f02151a516e7b76f2f

Thanks for helping,
Sr.

**Comment by Galexrt** [comment]
@reyman Can you try again with the new `dev` version of the image?
Please post your configs then again, so I can see what causes the LDAP config troubles.

**Comment by reyman** [comment]
Hi,
I cannot run the dev image, i have an error because of my password which contain `@`
```

zulip_1     |     AUTH_LDAP_BIND_PASSWORD = @xxxx@                                                                                                                                        [615/3401]
zulip_1     |                               ^
zulip_1     | SyntaxError: invalid syntax
zulip_1     | Traceback (most recent call last):
zulip_1     |   File "/home/zulip/deployments/current/manage.py", line 17, in <module>
zulip_1     |     " ".join(sys.argv)])
zulip_1     |   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
zulip_1     |     raise CalledProcessError(retcode, cmd)
zulip_1     | subprocess.CalledProcessError: Command '['/home/zulip/deployments/current/bin/log-management-command', '/home/zulip/deployments/current/manage.py checkconfig']' returned non-zero exit s
tatus 1

```

**Comment by Galexrt** [comment]
@reyman Could you check in what config file the AUTH_LDAP_BIND_PASSWORD with your password is located in.

**Comment by reyman** [comment]
@Galexrt Into `zulip_zulip_1:/home/zulip/deployments/current/zproject/settings.py`

**Comment by Galexrt** [comment]
@reyman I just pushed a fix that fixes the other fix that should have fixed that.
TL;DR This should fix the LDAP configuration problem.

Image should be built in about 15 minutes.

**Comment by reyman** [comment]
Same with EMAIL : 
```
zulip_1     | + su zulip -c '/home/zulip/deployments/current/manage.py checkconfig'
zulip_1     | Setting key "ADMIN_DOMAIN", type "array" in file "/etc/zulip/settings.py".
zulip_1     | Traceback (most recent call last):
zulip_1     |   File "/home/zulip/deployments/current/bin/log-management-command", line 10, in <module>
zulip_1     |     file_handler = logging.FileHandler(settings.MANAGEMENT_LOG_PATH)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/conf/__init__.py", line 48, in __getattr__
zulip_1     |     self._setup(name)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/conf/__init__.py", line 44, in _setup
zulip_1     |     self._wrapped = Settings(settings_module)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/conf/__init__.py", line 92, in __init__
zulip_1     |     mod = importlib.import_module(self.SETTINGS_MODULE)
zulip_1     |   File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
zulip_1     |     __import__(name)
zulip_1     |   File "/home/zulip/deployments/current/bin/../zproject/settings.py", line 975
zulip_1     |     EMAIL_HOST_USER = reyman64@gmail.com
zulip_1     |                               ^
zulip_1     | SyntaxError: invalid syntax
zulip_1     | Traceback (most recent call last):
zulip_1     |   File "/home/zulip/deployments/current/manage.py", line 17, in <module>
zulip_1     |     " ".join(sys.argv)])
zulip_1     |   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
zulip_1     |     raise CalledProcessError(retcode, cmd)
zulip_1     | subprocess.CalledProcessError: Command '['/home/zulip/deployments/current/bin/log-management-command', '/home/zulip/deployments/current/manage.py checkconfig']' returned non-zero exit s
tatus 1
zulip_zulip_1 exited with code 1
```

**Comment by Galexrt** [comment]
@reyman I pushed another fix. Repull the image and try again please.

**Comment by reyman** [comment]
OK, image run, but LDAP login page is not good :

**Comment by Galexrt** [comment]
@reyman Now that the image is running again, could you please post the new configs again, so I can see if the setting variables are at the right position.

**Comment by reyman** [comment]
Ok :)

https://gist.github.com/cf59c7ca4a03834029bb

**Comment by Galexrt** [comment]
@reyman you did forgot to remove a password in the `/etc/zulip/settings.py`. Please remove your password.

**Comment by reyman** [comment]
https://gist.github.com/2435a9163f1eaec6aa36

https://gist.github.com/9cf4cdf149e8438a9ff7

Thx, i remove the pass ...

**Comment by Galexrt** [comment]
The password is still there: https://gist.github.com/9cf4cdf149e8438a9ff7
at the bottom of the file, line: 352.

**Comment by reyman** [comment]
Ok i understand, password appear two times because there is two time the line 

```
AUTH_LDAP_USER_SEARCH = LDAPSearch("ou=Users,dc=comnmodel,dc=org",ldap.SCOPE_SUBTREE,"(uid=%(user)s)")
AUTH_LDAP_BIND_PASSWORD = '@xxx@'
```
https://gist.github.com/d3a5f19df9fb30929471

**Comment by Galexrt** [comment]
@reyman Could you please run `docker-compose stop` and `docker-compose rm` to get a clean container/config and post these then.

**Comment by reyman** [comment]
I make that, but env variable continue to appear two time in settings.py : 
```

➜  temp gist settings.py 
https://gist.github.com/6ce44961d67d2a6048e9
➜  temp gist settings_deployments.py 
https://gist.github.com/10f718f270e03283f9c0

```

**Comment by Galexrt** [comment]
I'm looking into the settings, that get written twice in the configs.

**Comment by Galexrt** [comment]
@reyman What environment variable do you use to set the password for LDAP authentication in the Zulip container?

**Comment by Galexrt** [comment]
@reyman I just tested the image myself and I don't get doubled settings in the configuration files.

Could you please write the whole process of how you start the containers, so I can understand from where the doubled settings originate.

**Comment by reyman** [comment]
Most of the time i make : 

- docker stop zulip_zulip_1
- docker rm zulip_zulip_1
- docker rmi 03204030fe #image of zulip_zulip_1
- docker pull galexrt/zulip:dev

I go to my folder `home/reyman/docker-compose-project/zulip` which contain the `docker-compose.yml` file   , then i run `docker-compose up -d`

After that, i make a docker cp to get the python settings file : 

- docker cp zulip_zulip_1:/etc/zulip/settings.py . 
- docker cp zulip_zulip_1:/home/zulip/deployments/current/zproject/settings.py settings_deployments.py

**Comment by Galexrt** [comment]
@reyman You should use `docker-compose stop`, to stop the whole Zulip
instance and then run `docker-compose up -d` to start it up again.

I strongly discourage from stopping and removing only the Zulip container
and then re-running `docker-compose up -d` to start the Zulip container
again.
It seems that this could may cause the doubled settings that are occuring
in your Zulip instance.

Please try with the following commands, to cycle the containers:
```
docker-compose stop
docker-compose up -d
```
After you cycled the containers with the commands, please check if the
settings are still doubled in the configuration files.

**Comment by reyman** [comment]
I try the two command, but i continue to have twice the config in the settings.py

**Comment by Galexrt** [comment]
@reyman It seems that only you are affected by the settings twice (I tested it and don't get doubled settings).
The doubled settings shouldn't cause problems with the LDAP authentication.

Could you try login into Zulip and post the logs from `/var/log/zulip/` as a gist, so I can see if it's a problem with the Docker image, a configuration issue or a Zulip problem.

**Comment by reyman** [comment]
@Galexrt I totally believe you, i try now to remove all docker images, redownload and relaunch using only the two command you give me. 

Then i post `/var/log/zulip` here when it's done.

**Comment by reyman** [comment]
@Galexrt i don't see anything in log in first view, is there a more specific ldap log ?

**Comment by Galexrt** [comment]
@reyman Sry that I have to say that, but I have no idea why the LDAP authentication isn't working (even the logs don't show any errors to me).

I would ask you to create a ticket at the zulip/zulip project and ask there if your LDAP authentication is correct and what this might cause then, because I have to admit that I've run out of ideas.

**Comment by reyman** [comment]
Hum, i think i found an explanation of wrong front end here : 
https://github.com/zulip/zulip/issues/210

Now i need to understand how can i capture correctly email in this frontend to test with email in ldap.

**Comment by Galexrt** [comment]
@reyman Thanks for digging around and findig this issue in the zulip/zulip project.

Do you still need my help configuring LDAP auth or is it not working cause of the bug?

**Comment by reyman** [comment]
Hi @Galexrt , Tim answer me, so i can make a new test later in the week-end.

I update the post in consequence, but i hope this is the last obstacle :)

**Comment by Galexrt** [comment]
@reyman That's good to hear! :)

**Comment by reyman** [comment]
Hi,

I try to run the special command to query_ldap with manage.py, but it's not easy because the script only run if it not launched with sudo. But zulip is into sudo folder.

Don't understand why, but the script return an error :(

```
root@3ea9e2c01c89:~/zulip# sudo -u zulip bash -c './manage.py query_ldap reyman64@gmail.com'
Traceback (most recent call last):
  File "./manage.py", line 24, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/lib/python2.7/dist-packages/django/core/management/__init__.py", line 338, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python2.7/dist-packages/django/core/management/__init__.py", line 330, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python2.7/dist-packages/django/core/management/base.py", line 393, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python2.7/dist-packages/django/core/management/base.py", line 444, in execute
    output = self.handle(*args, **options)
  File "/home/zulip/deployments/2016-03-15-15-45-30/zerver/management/commands/query_ldap.py", line 31, in handle
    query_ldap(**options)
  File "/home/zulip/deployments/2016-03-15-15-45-30/zerver/management/commands/query_ldap.py", line 18, in query_ldap
    ldap_attrs = _LDAPUser(backend, backend.django_to_ldap_username(email)).attrs
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 418, in _get_user_attrs
    self._load_user_attrs()
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 458, in _load_user_attrs
    if self.dn is not None:
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 411, in _get_user_dn
    self._load_user_dn()
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 474, in _load_user_dn
    self._search_for_user_dn()
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 494, in _search_for_user_dn
    results = search.execute(self.connection, {'user': self._username})
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 433, in _get_bound_connection
    self._bind()
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 718, in _bind
    sticky=True)
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 729, in _bind_as
    self._get_connection().simple_bind_s(force_str(bind_dn),
  File "/usr/lib/python2.7/dist-packages/django_auth_ldap/backend.py", line 743, in _get_connection
    self._connection = self.backend.ldap.initialize(uri)
  File "/usr/lib/python2.7/dist-packages/ldap/functions.py", line 91, in initialize
    return LDAPObject(uri,trace_level,trace_file,trace_stack_limit)
  File "/usr/lib/python2.7/dist-packages/ldap/ldapobject.py", line 70, in __init__
    self._l = ldap.functions._ldap_function_call(ldap._ldap_module_lock,_ldap.initialize,uri)
  File "/usr/lib/python2.7/dist-packages/ldap/functions.py", line 63, in _ldap_function_call
    result = func(*args,**kwargs)
ldap.LDAPError: (2, 'No such file or directory')
```

**Comment by reyman** [comment]
OK i found the problem, uri need to be indicated like this : 

`SETTING_AUTH_LDAP_SERVER_URI: "ldap://51.255.47.40:389"
`

When i test : 

```
root@c37c05dc0882:~/zulip# sudo -u zulip bash -c './manage.py query_ldap reyman64@gmail.com'
full_name: [u'Rey Sebastien']
```

Ldap connection seems to work in command line, but when i try directly with front end, that continue to don't work grrr ...

**Comment by Galexrt** [comment]
@reyman I have some good news for you :) Emojis are now working in the dev branch.

**Comment by Galexrt** [comment]
@reyman Do you set the `LDAP_APPEND_DOMAIN`? Could you please grep the config files for the variable name `LDAP_APPEND_DOMAIN`, to see if it's set somewhere.

When you can find a user through `manage.py`, there seems to be a problem with:
A) Zulip frontend, querying for something other than the email address (Maybe the related to `LDAP_APPEND_DOMAIN`)
B) Error in the Matrix

**Comment by reyman** [comment]
@Galexrt I set `LDAP_APPEND_DOMAIN` to None in docker-compose.yml

I check in `settings.py`, and i see something strange : 

```
# Specify the search base and the property to filter on that corresponds to the
# username.
AUTH_LDAP_USER_SEARCH = LDAPSearch("ou=users,dc=example,dc=com",
    ldap.SCOPE_SUBTREE, "(uid=%(user)s)")

# If the value of a user's "uid" (or similar) property is not their email
# address, specify the domain to append here.
LDAP_APPEND_DOMAIN = None

```
The AUTH_LDAP_USER_SEARCH don't correspond with docker-compose.yml : 

`SETTING_AUTH_LDAP_USER_SEARCH: "LDAPSearch(\"ou=Users,dc=comnmodel,dc=org\",ldap.SCOPE_SUBTREE,\"(mail=%(user)s)\")"`

**Comment by Galexrt** [comment]
@reyman Can you check the bottom of the `settings.py`? All settings are added to the bottom of the file.

**Comment by reyman** [comment]
@Galexrt You're right, everything seems correct .. i'm lost.

```
EXTERNAL_HOST = 'team.comnmodel.org'
AUTH_LDAP_BIND_DN = 'cn=admin,dc=comnmodel,dc=org'
NOREPLY_EMAIL_ADDRESS = 'noreply@comnmodel.org'
AUTH_LDAP_SERVER_URI = 'ldap://51.255.47.40:389'
ZULIP_ADMINISTRATOR = 'reyman64@gmail.com'
REDIS_HOST = 'redis'
AUTH_LDAP_USER_SEARCH = LDAPSearch("ou=Users,dc=comnmodel,dc=org",ldap.SCOPE_SUBTREE,"(mail=%(user)s)")
AUTH_LDAP_BIND_PASSWORD = 'xxxx'
RABBITMQ_HOST = 'rabbitmq'
DEFAULT_FROM_EMAIL = 'Zulip <noreply@comnmodel.org>'
MEMCACHED_LOCATION = 'memcached:11211'
AUTH_LDAP_APPEND_DOMAIN = 'None'
ADMIN_DOMAIN = 'comnmodel.org'
```

**Comment by Galexrt** [comment]
@reyman I know what the problem is :) The entrypoint.sh that sets the configuration should make the line look like this: `AUTH_LDAP_APPEND_DOMAIN = None`, without the quotes. I'm going to fix this.

**Comment by Galexrt** [comment]
@reyman **Important**: The variable is named `LDAP_APPEND_DOMAIN` and not `AUTH_LDAP_APPEND_DOMAIN`. Can you try again with that change?

**Comment by reyman** [comment]
@Galexrt I try with this change, with no amelioration :/

```
  SETTING_AUTH_LDAP_SERVER_URI: "ldap://51.255.47.40:389"
    SETTING_AUTH_LDAP_BIND_DN: "cn=admin,dc=comnmodel,dc=org"
    SETTING_AUTH_LDAP_BIND_PASSWORD: "xxx"
    SETTING_AUTH_LDAP_USER_SEARCH: "LDAPSearch(\"ou=Users,dc=comnmodel,dc=org\",ldap.SCOPE_SUBTREE,\"(mail=%(user)s)\")"
    SETTING_LDAP_APPEND_DOMAIN: "None"
```

**Comment by Galexrt** [comment]
@reyman The bottom of the `settings.py` still prints `LDAP_APPEND_DOMAIN: 'None'` with the quotes?

**Comment by reyman** [comment]
@Galexrt Nop !

```
AUTHENTICATION_BACKENDS = ('zproject.backends.ZulipLDAPAuthBackend',)
EXTERNAL_HOST = 'team.comnmodel.org'
AUTH_LDAP_BIND_DN = 'cn=admin,dc=comnmodel,dc=org'
NOREPLY_EMAIL_ADDRESS = 'noreply@comnmodel.org'
AUTH_LDAP_SERVER_URI = 'ldap://51.255.47.40:389'
ZULIP_ADMINISTRATOR = 'reyman64@gmail.com'
REDIS_HOST = 'redis'
AUTH_LDAP_USER_SEARCH = LDAPSearch("ou=Users,dc=comnmodel,dc=org",ldap.SCOPE_SUBTREE,"(mail=%(user)s)")
AUTH_LDAP_BIND_PASSWORD = 'xxx'
LDAP_APPEND_DOMAIN = None
RABBITMQ_HOST = 'rabbitmq'
DEFAULT_FROM_EMAIL = 'Zulip <noreply@comnmodel.org>'
MEMCACHED_LOCATION = 'memcached:11211'
ADMIN_DOMAIN = 'comnmodel.org'
```

**Comment by Galexrt** [comment]
@reyman The `LDAP_APPEND_DOMAIN` is added correctly without quotes. So it's another problem. 

Does `./manage.py query_ldap ...` still return the correct result?

**Comment by reyman** [comment]
@Galexrt yes ... 

```
root@1f4e35ec7e6d:~/zulip# sudo -u zulip bash -c './manage.py query_ldap reyman64@gmail.com'
full_name: [u'Rey Sebastien']
```

**Comment by reyman** [comment]
@Galexrt It seems it's linked to some special configuration during install : 
https://github.com/zulip/zulip/issues/315

**Comment by Galexrt** [comment]
@reyman You need to have a realm and the user in the realm to be able to log in. So at least it is now bug in the image.
Do you need some help on how to create a realm?

**Comment by reyman** [comment]
@Galexrt Hum, this is strange, when i try to create a realm, it say me that realm already exist ...

```
root@1f4e35ec7e6d:~/zulip# sudo -u zulip bash -c './manage.py create_realm --domain=comnmodel.org --name="com and model org"'
comnmodel.org already exists.
```

**Comment by Galexrt** [comment]
@reyman Could you ask at the zulip/zulip for help about LDAP authentication?

Sry, but I never configured or used LDAP authentication yet and can't help you much with it.

**Comment by reyman** [comment]
I create a new issue here.
I see one more time i have a double config, strange because i only use docker-compose stop and up command now. 

https://github.com/zulip/zulip/issues/591

**Comment by Galexrt** [comment]
@reyman I have no idea, why you get doubled settings. But the doubled settings shouldn't cause any problems. I'm closing this for now, if you get new information from the issue at zulip/zulip. Please reopen and mention me. Thanks!
